source/server/README.md

```diff
@@ -68,15 +75,15 @@ admin:
 
 ## Response Options config
 
-The ResponseOptions proto can be used in the test-server filter config or passed in `x-nighthawk-test-server-config``
-request header.
+The [ResponseOptions proto](/api/server/response_options.proto) can be used in the test-server
+and dynamic-delay filter config or passed in `x-nighthawk-test-server-config`` request header.
```

Comment:
This is a real good point, and I'm still mulling over how to best improve this. 

Some context: this is an artifact of my first stab at implementing this, where I used a single extension and injected it in two places. For technical reasons, that failed, I ran into trouble with what seemed Envoy getting confused about it - or I may have gotten confused myself as it complicated the extension more then I had hoped for. As we rely on the stock fault filter extension today for actually inducing latency, we need to be able to insert ourselves into two places in the filter pipeline: before and after the fault filter. Imho, ideally, for the end user, a response can be specified as a single entity, and how fragments of that apply to various parts within the test server (and its extensions) are implementation details. Right now we're sitting somewhere in the middle: request unifies it, but the static/file based configuration splits it, as more or less imposed by Envoy. 
I rewrote the docs part for a bit [1], but I'd love to see if we can improve the static/file-based configuration approach; yet I need to think some more about how (suggestions welcome!). Let me get back to this after thinking about it some more.

[1] 
```markdown
## Response Options config

The [ResponseOptions proto](/api/server/response_options.proto) is shared by
the `Test Server` and `Dynamic Delay` filter extensions. Each filter will
interpret the parts that are relevant to it. This allows specifying a what
a response should look like in a single message, which can be done at request
time via the optional `x-nighthawk-test-server-config` request-header.

### Test Server

- `response_body_size` - number of 'a' characters repeated in the response body.
- `response_headers` - list of headers to add to response. If `append` is set to
  `true`, then the header is appended.
- `echo_request_headers` - if set to `true`, then append the dump of request headers to the response
  body.

The response options above could be used to test and debug proxy or server configuration, for example, to verify request headers that are added by intermediate proxy:

<snip example req/resp>

This example shows that intermediate proxy has added `x-forwarded-proto` and
`x-forwarded-for` request headers.

### Dynamic Delay

The Dynamic Delay interprets the `oneof_delay_options` part in the [ResponseOptions proto](/api/server/response_options.proto). If specified, it can be used to:

- Configure a static delay
- Configure delays which linearly increase as the number of active requests grows, representing a simplified model of an overloaded server.
```

Reply:
Looking at the fault filter, the actual code that handles the delay injecting can be minimally expressed as:

```c++
// Set up delay
Http::FilterHeadersStatus FaultFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
  absl::optional<std::chrono::milliseconds> duration = delayDuration(headers);
  if (duration.has_value()) {
    delay_timer_ = decoder_callbacks_->dispatcher().createTimer(
        [this, &headers]() -> void { postDelayInjection(headers); });
    delay_timer_->enableTimer(duration.value(), &decoder_callbacks_->scope());
    return Http::FilterHeadersStatus::StopIteration;
  }
}

void FaultFilter::postDelayInjection(const Http::RequestHeaderMap& request_headers) {
  resetTimerState();
  decoder_callbacks_->continueDecoding();
}

```

We could consider taking ownership of that by copying it. 
Potential advantages: 

- we can inject into a single place
- easier setup for the end-user, single config/extension
- Have a relatively easy and surefire path to generically map/route qs params and/or headers to proto config fields
- It may also allow us to switch to sub-ms granularity with a small tweak

Possible disadvantages:

- We own that code
- Need to add unit tests to support it, can no longer rely on Envoy for that
- Fault filter may still need to be pulled in for other functionality (fault injection)
- Looking at the code, the fault filter is more flexible in configuration, in that it supports route-level
configuration and maybe more

Reply:
Another idea: could we use the fault filter as a base and derive from it?

Reply:
I think if the code is as simple as you point to, there's no good reason to be doing a lot of sharing with the fault injection filter. If you want to more generally extend it, then you could consider this pattern. Keep in mind that subclassing may expose your code to fragility, since there is no hard contract on the interface in the Envoy repo.

Reply:
I looked some more into this, and the fault filter has evolved for quite a bit since I last looked at it, and learned new cool tricks.

Also, I made a small trip to explore taking the fault filter as a base class;  [Here's a diff](https://github.com/oschaaf/nighthawk/compare/test-server-dynamic-delay...oschaaf:explore-fault-filter-as-a-base?expand=1), but a high level description of the code-level contact surface would be:

- The constructor -- we take NHs own `ResponseOptions` proto, and translate that into a minimal configuration for the fault filter base class which it takes in its constructor to set it up like the way we need it. 
- We override decodeHeaders() and synthesize the request header that the fault filter understands to trigger the actual delay and forward the call to the base via  `Envoy::Extensions::HttpFilters::Fault::FaultFilter::decodeHeaders(headers, end_stream)` when our own override is done, which is the only difference with what we have right now.
- We override `HttpDynamicDelayDecoderFilter::setDecoderFilterCallbacks()` to grab a copy of the callbacks which we'll only use in case we didn't understand the per-request configuration to short circuit and send out an error response. We forward the call to the base, obviously, after grabbing the copy. This would be relatively fragile, as the Fault filter may not anticipate us directly replying. We could instead not do that and log an error about bad configs to make it more robust. We no longer need to override `setDecoderFilterCallbacks()`.

Challenges:

- Includes aren't routed through external/envoy (see code level todo in the diff above), requires an upstream change in the build system to make that happen (probably similar to what we did earlier).
- Need to check the stats_prefix_ argument that the fault filter takes & verify counter behaviour is allright.

I feel similar in that imho it depends on how we envision the test server functionality to expand in the future. For example, one of the fault filters new tricks (to me at least) is throttling streams; I wouldn't be surprised if similar functionality is going to be requested at some point for Nighthawk's test server. If we want to extend / modify behavior for that as well, using the fault filter as a base might be convenient -- iff it turns out fragility isn't too bad. Otoh, if the fault filter moves in a direction that would enforce us to re-implement some things, it's good take we do take ownership of the proto api, as that limits impact to the end-user: The fault filter as a base becomes an implementation detail.

@dubious90 / @mum4k do you have any thoughts on this?

Reply:
also cc @jmarantz ^^

Reply:
My understanding of this is limited, so I'm going to defer to you, Harvey, and Jakub on what makes sense here.

Reply:
Thank you for the detailed analysis here @oschaaf, this is very useful. I would suggest we err on the side of minimizing maintenance cost and code churn and implement this without the inheritance, keeping ourselves to the external/envoy imports.

Correct me if I am wrong, but I believe we don't have use cases for the test server that would require us to modify the behavior of the fault filter. Even if we find some, it might be more beneficial to push such modifications upstream by adding options or improving the fault filter itself. Once we find a use case that is very specific to Nighthawk, we can reopen this discussion.

Reply:
So I’d be happy to back out the basing-off of the fault filter that I merged earlier today, and switch this  to copy the relevant code so we own it, see https://github.com/envoyproxy/nighthawk/pull/390#discussion_r450503227

That takes the fault filter out of the equation, and allows us to provide our own configuration,  but also has a side effect: while right now  we just need to prove we wire up correctly to the fault filter, and can  rely on Envoy for testing that that does what it says on the tin with respect to delays , now we’d also need to test that ourselves to do it right.

IMHO upstreaming might not *always* be an option, as goals are not guaranteed to align; which is why we are here I guess? :-) so maybe copying the delay code is a middle ground then?